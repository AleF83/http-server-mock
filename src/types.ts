import { v4 as uuidV4 } from 'uuid';

export interface ServiceMockOptions {
  name: string;
  port: number;
}

export interface ServiceMockInfo {
  id: string;
  name: string;
  port: number;
}

export class ServiceMock {
  private _id: string;
  private _name: string;
  private _port: number;

  constructor(options: ServiceMockOptions) {
    this._id = uuidV4();
    this._name = options.name;
    this._port = options.port;
  }

  get id(): string {
    return this._id;
  }

  get name(): string {
    return this._name;
  }

  get port(): number {
    return this._port;
  }

  get info(): ServiceMockInfo {
    return {
      id: this.id,
      name: this.name,
      port: this.port,
    };
  }
}
