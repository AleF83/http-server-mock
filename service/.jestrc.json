{
  "projects": [
    {
      "displayName": "Service - Unit Tests",
      "collectCoverage": true,
      "collectCoverageFrom": ["<rootDir>/src/**/!(*.spec).js"],
      "preset": "ts-jest",
      "rootDir": ".",
      "testEnvironment": "node",
      "testMatch": ["<rootDir>/src/**/*.spec.ts"],
      "testPathIgnorePatterns": ["<rootDir>/dist/", "<rootDir>/node_modules/"],
      "verbose": true
    },
    {
      "displayName": "Service - Integration Tests",
      "preset": "ts-jest",
      "rootDir": ".",
      "testEnvironment": "node",
      "testMatch": ["<rootDir>/tests/**/*.spec.ts"],
      "testPathIgnorePatterns": ["<rootDir>/dist/", "<rootDir>/node_modules/"],
      "verbose": true
    }
  ],
  "preset": "ts-jest",
  "rootDir": ".",
  "testEnvironment": "node",
  "testMatch": ["**/?(*.)+(spec|e2e-spec).ts"],
  "testPathIgnorePatterns": ["<rootDir>/dist/", "<rootDir>/node_modules/"],
  "verbose": true
}
