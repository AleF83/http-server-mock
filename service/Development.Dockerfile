# syntax = docker/dockerfile:experimental
FROM node:lts-buster-slim

RUN --mount=type=cache,target=/usr/local/share/.cache/yarn/v6,sharing=locked yarn global add lerna

WORKDIR /http-server-mock

COPY package.json yarn.lock lerna.json ./

COPY ./service/package.json ./service/tsconfig.json ./service/
COPY ./packages ./packages
RUN --mount=type=cache,target=/usr/local/share/.cache/yarn/v6,sharing=locked lerna bootstrap
COPY ./service/package.json ./service
COPY ./service/src ./service/src

RUN lerna run build

ENV NODE_ENV=development

WORKDIR /http-server-mock/service

CMD [ "yarn", "debug" ]
