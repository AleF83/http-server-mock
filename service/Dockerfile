# syntax = docker/dockerfile:experimental
FROM node:lts-buster-slim AS base

FROM base AS dependencies
WORKDIR /app
COPY package.json yarn.lock ./
RUN --mount=type=cache,target=/usr/local/share/.cache/yarn/v6,sharing=locked yarn --pure-lockfile --production true

FROM base AS build
WORKDIR /app
COPY package.json yarn.lock ./
RUN --mount=type=cache,target=/usr/local/share/.cache/yarn/v6,sharing=locked yarn --pure-lockfile --production false
COPY tsconfig.json tsconfig.json .jestrc.json ./
COPY src ./src
COPY tests ./tests
RUN yarn test:unit && yarn test:integration
RUN yarn build

FROM base AS release
WORKDIR /app
COPY --from=dependencies /app/node_modules ./node_modules
COPY --from=build /app/dist ./dist
COPY package.json ./

ENV NODE_ENV=production

CMD [ "yarn", "start" ]

