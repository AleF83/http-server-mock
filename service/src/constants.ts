export const ERR_MOCK_PORT_ALREADY_IN_USE = 'The port is already in use';
export const ERR_MOCK_ID_DOES_NOT_EXIST = "The mock with giver id doesn't exist";
