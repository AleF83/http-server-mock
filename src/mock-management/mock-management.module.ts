import { Module } from '@nestjs/common';
import { MockManagementController } from './mock-management.controller';
import { MockManagementService } from './mock-management.service';

@Module({
  imports: [],
  controllers: [MockManagementController],
  providers: [MockManagementService],
})
export class MockManagementModule {}
