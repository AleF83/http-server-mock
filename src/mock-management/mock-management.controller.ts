import {
  Controller,
  Get,
  Post,
  Delete,
  Body,
  Param,
  HttpStatus,
  HttpCode,
  BadRequestException,
  NotFoundException,
} from '@nestjs/common';
import * as Constants from '../constants';
import { ServiceMockInfo, ServiceMockOptions } from '../types';
import { MockManagementService } from './mock-management.service';

@Controller()
export class MockManagementController {
  constructor(private readonly _mockManager: MockManagementService) {}

  @Get()
  getMocks(): ServiceMockInfo[] {
    return this._mockManager.getMocks().map((sm) => sm.info);
  }

  @Get('/:id')
  getMock(@Param('id') id: string): ServiceMockInfo {
    const mock = this._mockManager.getMock(id);
    if (!mock) {
      throw new NotFoundException(Constants.ERR_MOCK_ID_DOES_NOT_EXIST);
    }
    return mock.info;
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  createMock(@Body() options: ServiceMockOptions): { id: string } {
    try {
      const id = this._mockManager.createMock(options);
      return { id };
    } catch (error) {
      if (error.message === Constants.ERR_MOCK_PORT_ALREADY_IN_USE) {
        throw new BadRequestException(Constants.ERR_MOCK_PORT_ALREADY_IN_USE);
      }
      throw error;
    }
  }

  @Delete('/:id')
  deleteMock(@Param('id') id: string): ServiceMockInfo | undefined {
    return this._mockManager.deleteMock(id)?.info;
  }
}
