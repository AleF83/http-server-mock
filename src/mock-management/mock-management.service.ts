import { Injectable } from '@nestjs/common';
import { Dictionary } from 'typescript-collections';
import * as Constants from '../constants';
import { ServiceMock, ServiceMockOptions } from '../types';

@Injectable()
export class MockManagementService {
  private _mocks: Dictionary<string, ServiceMock>;
  private _mocksByPort: Dictionary<number, ServiceMock>;

  constructor() {
    this._mocks = new Dictionary<string, ServiceMock>();
    this._mocksByPort = new Dictionary<number, ServiceMock>();
  }

  getMocks(): ServiceMock[] {
    return this._mocks.values();
  }

  getMock(id: string): ServiceMock | undefined {
    return this._mocks.getValue(id);
  }

  createMock(options: ServiceMockOptions): string {
    if (this._mocksByPort.containsKey(options.port)) {
      throw new Error(Constants.ERR_MOCK_PORT_ALREADY_IN_USE);
    }
    const mock = new ServiceMock(options);
    this._mocks.setValue(mock.id, mock);
    this._mocksByPort.setValue(mock.port, mock);
    return mock.id;
  }

  deleteMock(id: string): ServiceMock | undefined {
    const mock = this._mocks.getValue(id);
    if (!mock) {
      return;
    }
    this._mocksByPort.remove(mock.port);
    return this._mocks.remove(id);
  }
}
