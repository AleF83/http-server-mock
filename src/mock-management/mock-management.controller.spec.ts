import { Test, TestingModule } from '@nestjs/testing';
import { v4 as uuidV4 } from 'uuid';
import * as Constants from '../constants';
import { ServiceMockOptions, ServiceMock } from '../types';
import { MockManagementController } from './mock-management.controller';
import { MockManagementService } from './mock-management.service';

const createTestServiceMockOptions = (): ServiceMockOptions => ({
  name: 'my service mock',
  port: 3333,
});

const createTestServiceMock = (): ServiceMock => new ServiceMock(createTestServiceMockOptions());

describe('Mock Management Controller', () => {
  let controller: MockManagementController;
  let service: MockManagementService;

  beforeEach(async () => {
    const app: TestingModule = await Test.createTestingModule({
      controllers: [MockManagementController],
      providers: [MockManagementService],
    }).compile();

    controller = app.get(MockManagementController);
    service = app.get(MockManagementService);
  });

  afterEach(() => {
    jest.restoreAllMocks();
    jest.resetAllMocks();
  });

  test('Get mocks', () => {
    // Arrange
    const testMock = createTestServiceMock();
    jest.spyOn(service, 'getMocks').mockReturnValueOnce([testMock]);

    // Act
    const mocks = controller.getMocks();

    // Assert
    expect(mocks).toHaveLength(1);
    expect(mocks[0]).toMatchSnapshot({ id: expect.any(String) });
  });

  test('Get mock', () => {
    // Arrange
    const testMock = createTestServiceMock();
    jest.spyOn(service, 'getMock').mockReturnValueOnce(testMock);

    // Act
    const mock = controller.getMock(testMock.id);

    // Assert
    expect(mock).toMatchSnapshot({ id: expect.any(String) });
  });

  test('Get nonexistent mock', () => {
    // Arrange
    const mockId = uuidV4();
    // eslint-disable-next-line unicorn/no-useless-undefined
    jest.spyOn(service, 'getMock').mockReturnValueOnce(undefined);

    // Act
    const shouldThrow = () => controller.getMock(mockId);

    // Assert
    expect(shouldThrow).toThrow(Constants.ERR_MOCK_ID_DOES_NOT_EXIST);
  });

  test('Create mock', () => {
    // Arrange
    const mockOptions = createTestServiceMockOptions();
    const testMock = new ServiceMock(mockOptions);
    jest.spyOn(service, 'createMock').mockReturnValueOnce(testMock.id);

    // Act
    const mockId = controller.createMock(mockOptions);

    // Assert
    expect(mockId).toEqual({ id: testMock.id });
  });

  test('Create mock with port that already is in use', () => {
    // Arrange
    const mockOptions = createTestServiceMockOptions();
    jest.spyOn(service, 'createMock').mockImplementationOnce(() => {
      throw new Error(Constants.ERR_MOCK_PORT_ALREADY_IN_USE);
    });

    // Act
    const shouldThrow = () => controller.createMock(mockOptions);

    // Assert
    expect(shouldThrow).toThrow(Constants.ERR_MOCK_PORT_ALREADY_IN_USE);
  });

  test('Create mock throws internal error', () => {
    // Arrange
    const mockOptions = createTestServiceMockOptions();
    const errorMessage = 'Something wrong';
    jest.spyOn(service, 'createMock').mockImplementationOnce(() => {
      throw new Error(errorMessage);
    });

    // Act
    const shouldThrow = () => controller.createMock(mockOptions);

    // Assert
    expect(shouldThrow).toThrow(errorMessage);
  });

  test('Delete mock', () => {
    // Arrange
    const testMock = createTestServiceMock();
    jest.spyOn(service, 'deleteMock').mockReturnValueOnce(testMock);

    // Act
    const mock = controller.deleteMock(testMock.id);

    // Assert
    expect(mock?.id).toEqual(testMock.id);
  });

  test('Delete nonexisting mock', () => {
    // Arrange
    const mockId = uuidV4();
    // eslint-disable-next-line unicorn/no-useless-undefined
    jest.spyOn(service, 'deleteMock').mockReturnValueOnce(undefined);

    // Act
    const mock = controller.deleteMock(mockId);

    // Assert
    expect(mock).toBeUndefined();
  });
});
