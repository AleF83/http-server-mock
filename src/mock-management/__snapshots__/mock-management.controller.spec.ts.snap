// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mock Management Controller Get mock 1`] = `
Object {
  "id": Any<String>,
  "name": "my service mock",
  "port": 3333,
}
`;

exports[`Mock Management Controller Get mocks 1`] = `
Object {
  "id": Any<String>,
  "name": "my service mock",
  "port": 3333,
}
`;
